pkg/pgmodel/ingestor/trace/trace_batcher.go

```diff
@@ -0,0 +1,283 @@
+	inChs := make([]chan insertTracesReq, config.Batchers)
+	for i := range inChs {
+		inChs[i] = make(chan insertTracesReq, config.MaxBatchSize*3) // buffer for incoming requests, especially important for async acks
+	}
+	return &Batcher{
+		batchWriter:     newBatchWriter(config.Writers, writer, bufferedBatches),
+		in:              inChs,
+		stop:            make(chan struct{}),
+		bufferedBatches: bufferedBatches,
+		config:          config,
+	}
+}
+
+func (b *Batcher) send(req insertTracesReq, batcherIdx int) {
+	b.in[batcherIdx] <- req
+}
+
+func validateConfig(config *BatcherConfig) {
+	if config.Batchers == 0 {
+		config.Batchers = DefaultBatchWorkers
+	}
+	if config.MaxBatchSize == 0 {
+		config.MaxBatchSize = DefaultBatchSize
+	}
+	if config.BatchTimeout == 0 {
+		config.BatchTimeout = DefaultBatchTimeout
+	}
+	if config.Batchers < 1 || config.Writers < 1 {
+		panic("number of batchers and writeres must be greater then zero")
+	}
+}
+
+func (b *Batcher) Run() {
+	for i := 0; i < b.config.Batchers; i++ {
+		b.wg.Add(1)
+		go func(idx int) {
+			defer b.wg.Done()
+			b.batch(idx)
+		}(i)
+	}
+	b.batchWriter.run()
+}
+
+func (b *Batcher) batch(batchIdx int) {
+	_, batcherSpan := tracer.Default().Start(context.Background(), "create-batches")
+	ticker := time.NewTicker(b.config.BatchTimeout)
+	defer ticker.Stop()
+	batch := NewBatch(b.config.MaxBatchSize)
+	batcherSpan.AddEvent("New Batch")
+	flushBatch := func(batch *Batch) *Batch {
+		b.bufferedBatches <- *batch
+		batcherSpan.AddEvent("Batch sent to buffer")
+		batcherSpan.AddEvent("New Batch")
+		metrics.IngestorPendingBatches.With(prometheus.Labels{"type": "trace"}).Inc()
+		return NewBatch(b.config.MaxBatchSize)
+	}
+	processReq := func(req insertTracesReq) {
+		metrics.IngestorRequestsQueued.With(prometheus.Labels{"type": "trace", "queue_idx": fmt.Sprintf("%d", batchIdx)}).Dec()
+		batch.add(req)
+		if batch.addCounter == 1 {
+			// we reset timeout once we add first request into the batch
+			ticker.Reset(b.config.BatchTimeout)
+		}
+		if batch.isFull() {
+			metrics.IngestorBatchFlushTotal.With(prometheus.Labels{"type": "trace", "reason": "size"}).Inc()
+			batch = flushBatch(batch)
+		}
+	}
+	for {
+		select {
+		case item := <-b.in[batchIdx]:
+			processReq(item)
+		case <-ticker.C:
+			batcherSpan.AddEvent("Batch timeout reached")
+			metrics.IngestorBatchFlushTotal.With(prometheus.Labels{"type": "trace", "reason": "timeout"}).Inc()
+			if !batch.isEmpty() {
+				batch = flushBatch(batch)
+			}
+		case <-b.stop:
+			timeoutCtx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+			defer cancel()
+			close(b.in[batchIdx])
+			// shutting down. let's drain the requests
+		loop:
+			for {
+				select {
+				case item, ok := <-b.in[batchIdx]:
+					if !ok {
+						break loop
+					}
+					processReq(item)
+				case <-timeoutCtx.Done():
+					log.Warn("msg", "Forced batcher shutdown due to timeout. Some request might not be persisted.", "batcher", batchIdx)
```

Comment:
can we check len(b.in) to see if "Some request might not be persisted" and print that only if some requests were dropped?